vela-templates/definitions/internal/trait/gateway.cue

```diff
@@ -9,55 +9,64 @@ gateway: {
 
 		status: {
 			customStatus: #"""
-				if context.outputs.ingress.status.loadBalancer.ingress == _|_ {
+				let nameSuffix = {
+				  if parameter.name != _|_ { "-" + parameter.name }
+				  if parameter.name == _|_ { "" }
+				}
+				let ingressMetaName = context.name + nameSuffix
+				let ig  = [for i in context.outputs if (i.kind == "Ingress") && (i.metadata.name == ingressMetaName) {i}][0]
+				igs: *null | string
+				if ig != _|_ if ig.status != _|_ if ig.status.loadbalancer != _|_ {
+				  igs: ig.status.loadbalancer.ingress[0]
+				}
+				igr: *null | string
+				if ig != _|_ if ig.spec != _|_  {
+				  igr: ig.spec.rules[0]
+				}
+				if igs == _|_ {
 				  message: "No loadBalancer found, visiting by using 'vela port-forward " + context.appName + "'\n"
 				}
-				if context.outputs.ingress.status.loadBalancer.ingress != _|_ {
-					let igs = context.outputs.ingress.status.loadBalancer.ingress
-					let host = context.outputs.ingress.spec.rules[0].host
-				  if igs[0].ip != _|_ {
-				  	if host != _|_ {
-					    message: "Visiting URL: " + context.outputs.ingress.spec.rules[0].host + ", IP: " + igs[0].ip
-				  	}
-				  	if host == _|_ {
-					    message: "Host not specified, visit the cluster or load balancer in front of the cluster with IP: " + igs[0].ip
-				  	}
+				if igs != _|_ {
+				  if igs.ip != _|_ {
+				    if igr.host != _|_ {
+				      message: "Visiting URL: " + igr.host + ", IP: " + igs.ip + "\n"
+				    }
+				    if igr.host == _|_ {
+				      message: "Host not specified, visit the cluster or load balancer in front of the cluster, IP: " + igs.ip + "\n"
+				    }
 				  }
-				  if igs[0].ip == _|_ {
-				  	if host != _|_ {
-						  message: "Visiting URL: " + context.outputs.ingress.spec.rules[0].host
-						}
-				  	if host == _|_ {
-					    message: "Host not specified, visit the cluster or load balancer in front of the cluster"
-						}
+				  if igs.ip == _|_ {
+				    if igr.host != _|_ {
+				      message: "Visiting URL: " + igr.host + "\n"
+				    }
+				    if igs.host == _|_ {
+				      message: "Host not specified, visit the cluster or load balancer in front of the cluster\n"
+				    }
 				  }
 				}
 				"""#
 			healthPolicy: #"""
-				isHealth: len(context.outputs.service.spec.clusterIP) > 0
+				let nameSuffix = {
+				  if parameter.name != _|_ { "-" + parameter.name }
+				  if parameter.name == _|_ { "" }
+				}
+				let ingressMetaName = context.name + nameSuffix
+				let ig  = [for i in context.outputs if (i.kind == "Ingress") && (i.metadata.name == ingressMetaName) {i}][0]
+				let igstat = (ig.status.loadBalancer.ingress[0].ip != _|_)
+				isHealth: igstat
```

Comment:
Hi @Kolossi I just checked out the branch and checked the unit test. 

In this test case: "test application with trait-gateway support https protocol". It will test a app with gateway trait. and since you change the ingress healthPolicy to check `.status.loadBalancer.ingress` field of Ingress which is not generated in unit test environment. 

I add some code to check the process:

```go

		bs, _ := yaml.Marshal(curApp)
		fmt.Fprintf(GinkgoWriter, "curApp: \n%s\n", string(bs))

		ing := &networkingv1.Ingress{}
		err := k8sClient.Get(ctx, client.ObjectKey{Name: "myworker", Namespace: "vela-test-with-trait-gateway-https"}, ing)
		Expect(err).Should(BeNil())
		bs, _ = yaml.Marshal(ing)
		fmt.Fprintf(GinkgoWriter, "ingress: \n%s\n", string(bs))

```

You can run unit test locally to see details.
```shell
ginkgo -focus "test application with trait-gateway support https protocol" ./pkg/controller/core.oam.dev/v1alpha2/application
 ```

My suggestion here is to add a gateway trait in `testdata` and use that. Keep the healthPolicy simple and make it pass the workflow when there's no ingress controller.

Reply:
@Kolossi Yes. I think it's OK to leave the built-in gateway trait's behavior like that as the same before (Better because it'll check the Ingress.) 

Now I got the second requirements to link multiple domains with different certs to one service.

Reply:
OK, great I will do the unittest checks today.

> Now I got the second requirements to link multiple domains with different certs to one service.

This is the requirement that had me implement this, but I didn't quite get this comment.  Are you saying this needs work to achieve your second requirement, or you want this merged soon as it now meets a requirement you have? :-)

Reply:
The latter. :D It's a common case I think.

Reply:
I tried doing `make && make test` in my docker-based builder, but this failed in `./pkg/definition/gen_sdk/gen_skd_test.go` since `exec: "docker": executable file not found in $PATH`.  It's running _in_ docker, it doesn't _have_ docker :-)

Anyway I tried your suggested command `ginkgo -focus "test application with trait-gateway support https protocol" ./pkg/controller/core.oam.dev/v1alpha2/application` but this fails apparently with a ginko syntax error as the helptext is displayed:

```
root@cb45868a70c6:~/kubevela# ginkgo -focus "test application with trait-gateway support https protocol" ./pkg/controller/core.oam.dev/v1alpha2/application
flag provided but not defined: -ginkgo.timeout
Usage of /root/kubevela/pkg/controller/core.oam.dev/v1alpha2/application/application.test:
  -ginkgo.debug
        If set, ginkgo will emit node output to files when running in parallel.
  -ginkgo.dryRun
        If set, ginkgo will walk the test hierarchy without actually running anything.  Best paired with -v.
  -ginkgo.failFast
        If set, ginkgo will stop running a test suite after a failure occurs.
  -ginkgo.failOnPending
        If set, ginkgo will mark the test suite as failed if any specs are pending.
  -ginkgo.flakeAttempts int
        Make up to this many attempts to run each spec. Please note that if any of the attempts succeed, the suite will not be failed. But any failures will still be recorded. (default 1)
  -ginkgo.focus value
        If set, ginkgo will only run specs that match this regular expression. Can be specified multiple times, values are ORed.
  -ginkgo.noColor
        If set, suppress color output in default reporter.
  -ginkgo.noisyPendings
        If set, default reporter will shout about pending tests. (default true)
  -ginkgo.noisySkippings
        If set, default reporter will shout about skipping tests. (default true)
  -ginkgo.parallel.node int
        This worker node's (one-indexed) node number.  For running specs in parallel. (default 1)
  -ginkgo.parallel.streamhost string
        The address for the server that the running nodes should stream data to.
  -ginkgo.parallel.synchost string
        The address for the server that will synchronize the running nodes.
  -ginkgo.parallel.total int
        The total number of worker nodes.  For running specs in parallel. (default 1)
  -ginkgo.progress
        If set, ginkgo will emit progress information as each spec runs to the GinkgoWriter.
  -ginkgo.randomizeAllSpecs
        If set, ginkgo will randomize all specs together.  By default, ginkgo only randomizes the top level Describe, Context and When groups.
  -ginkgo.regexScansFilePath
        If set, ginkgo regex matching also will look at the file path (code location).
  -ginkgo.reportFile string
        Override the default reporter output file path.
  -ginkgo.reportPassed
        If set, default reporter prints out captured output of passed tests.
  -ginkgo.seed int
        The seed used to randomize the spec suite. (default 1683539470)
  -ginkgo.skip value
        If set, ginkgo will only run specs that do not match this regular expression. Can be specified multiple times, values are ORed.
  -ginkgo.skipMeasurements
        If set, ginkgo will skip any measurement specs.
  -ginkgo.slowSpecThreshold float
        (in seconds) Specs that take longer to run than this threshold are flagged as slow by the default reporter. (default 5)
  -ginkgo.succinct
        If set, default reporter prints out a very succinct report
  -ginkgo.trace
        If set, default reporter prints out the full stack trace when a failure occurs
  -ginkgo.v
        If set, default reporter print out all specs as they begin.
  -kubeconfig string
        Paths to a kubeconfig. Only required if out-of-cluster.
  -test.bench regexp
        run only benchmarks matching regexp
  -test.benchmem
        print memory allocations for benchmarks
  -test.benchtime d
        run each benchmark for duration d (default 1s)
  -test.blockprofile file
        write a goroutine blocking profile to file
  -test.blockprofilerate rate
        set blocking profile rate (see runtime.SetBlockProfileRate) (default 1)
  -test.count n
        run tests and benchmarks n times (default 1)
  -test.coverprofile file
        write a coverage profile to file
  -test.cpu list
        comma-separated list of cpu counts to run each test with
  -test.cpuprofile file
        write a cpu profile to file
  -test.failfast
        do not start new tests after the first test failure
  -test.fuzz regexp
        run the fuzz test matching regexp
  -test.fuzzcachedir string
        directory where interesting fuzzing inputs are stored (for use only by cmd/go)
  -test.fuzzminimizetime value
        time to spend minimizing a value after finding a failing input (default 1m0s)
  -test.fuzztime value
        time to spend fuzzing; default is to run indefinitely
  -test.fuzzworker
        coordinate with the parent process to fuzz random values (for use only by cmd/go)
  -test.list regexp
        list tests, examples, and benchmarks matching regexp then exit
  -test.memprofile file
        write an allocation profile to file
  -test.memprofilerate rate
        set memory allocation profiling rate (see runtime.MemProfileRate)
  -test.mutexprofile string
        write a mutex contention profile to the named file after execution
  -test.mutexprofilefraction int
        if >= 0, calls runtime.SetMutexProfileFraction() (default 1)
  -test.outputdir dir
        write profiles to dir
  -test.paniconexit0
        panic on call to os.Exit(0)
  -test.parallel n
        run at most n tests in parallel (default 8)
  -test.run regexp
        run only tests and examples matching regexp
  -test.short
        run smaller test suite to save time
  -test.shuffle string
        randomize the execution order of tests and benchmarks (default "off")
  -test.testlogfile file
        write test action log to file (for use only by cmd/go)
  -test.timeout d
        panic test binary after duration d (default 0, timeout disabled)
  -test.trace file
        write an execution trace to file
  -test.v
        verbose: print additional output

Ginkgo ran 1 suite in 8.1917425s

Test Suite Failed
```

I just don't have a dev setup ready to test all this I'm afraid.

I'll update the health as suggested in the previous comment and repush.

Reply:
1. There's a SDK test depends on Docker which is hard to remove. You can simply run a single test. And it takes so long to run the whole test.
2. Check the ginkgo version maybe? I'm using `Ginkgo Version 1.16.5`
3. Another question, have you tested this definition in the component with multiple gateway trait?

Reply:
> 2. Check the ginkgo version maybe? I'm using Ginkgo Version 1.16.5

The [instructions by wonderflow](https://github.com/wonderflow/kubevela.io/blob/b4b7bae0a90e0b087df79e5ba5c46fdca072e4f6/docs/contributor/code-contribute.md#run-kubevela-locally) have it installing v2 latest.  Mine picked up v2.9.2!!!

> 3. Another question, have you tested this definition in the component with multiple gateway trait?

Well we corrected a few things in this PR, so actually I haven't used this particular version. It does pass the tests for single usage so I'd expect it to be ok, but I'll test this at some point tomorrow. I'll post the dry-run result to demonstrate.